### PR TITLE
author/ sheets api

### PR DESCRIPTION
**Pull request by adirar111**
progress on #66

okaaaaaaaaaaaaaaaaaaaaaaaaaaay
i think this kind of works now (?)

none of the actual feature is implemented yet, but this PR adds an auth layer, a sheets API layer, and an author API layer

would appreciate if you could pull it down and try playing around with it in case there's anything i missed -- you'll need the api key and client id tho

#### asking for api keys

Screen Shot 2022-10-14 at 3 53 42 PM

#### the sheets data

Screen Shot 2022-10-14 at 4 09 27 PM

#### how the author api should work
```ts
> await Author.getAuthorTagsByUUID("achebechinua");

// returned
{
    "uuid": "achebechinua",
    "name": "chinua achebe",
    "tags": [
        "a",
        "a",
        "a",
        "a",
        "a"
    ],
    "sheetsCellIndex": "A2:C2"
}
```

```ts
> await Author.createNewAuthorWithTags("kincaidjamaica", "Jamaica Kincaid", [
		"Black author",
		"Femme author",
		"Carribean author",
	]);

// returned
1
```

```ts
// get all authors
> await Author.getAllAuthorTags();

// returned
[
    {
        "uuid": "achebechinua",
        "name": "chinua achebe",
        "tags": [
            "Black author",
            "Nigerian author",
            "Male author"
        ],
        "sheetsCellIndex": "A2:C2"
    },
    {
        "uuid": "kincaidjamaica",
        "name": "Jamaica Kincaid",
        "tags": [
            "Black author",
            "Femme author",
            "Carribean author"
        ],
        "sheetsCellIndex": "A3:C3"
    }
]
```

```ts
// updating a non-existent author
> await Author.updateAuthorTags("bob's son", [
		"you're listening to bob's son now",
		"play bob's son",
		"turn on bob's son",
	]);


// returned
0
```

### CODE REVIEW

File: extension/ts/objects/auth.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+};
+
+const clearAPICredentials = () => {
+	localStorage.removeItem(API_KEY_LOCAL_STORAGE);
+	localStorage.removeItem(CLIENT_ID_LOCAL_STORAGE);
+};
+
+const isAPICredentialsSavedToLocalStorage = () =>
+	!!localStorage[API_KEY_LOCAL_STORAGE] &&
+	localStorage[API_KEY_LOCAL_STORAGE] !== "null" &&
```
Comment (adirar111) [inline extension/ts/objects/auth.ts:45]: if you don't save anything or the auth flow gets into a weird state, sometimes the string value `"null"` gets saved to local storage

File: extension/ts/objects/auth.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		const input = document.createElement("input");
+		input.setAttribute("type", "hidden");
+		input.setAttribute("name", p);
+		input.setAttribute("value", params[p]);
+		form.appendChild(input);
+	}
+
+	// Add form to page and submit it to open the OAuth 2.0 endpoint.
+	document.body.appendChild(form);
+	form.submit();
```
Comment (adirar111) [inline extension/ts/objects/auth.ts:82]: this is all taken from the OAuth docs

File: extension/ts/objects/auth.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	const fragmentString = location.hash.substring(1);
+
+	// The token and its related parameters get appended to the redirected URL
+	// Parse query string to see if page request is coming from OAuth 2.0 server.
+	const params = {};
+	const regex = /([^&=]+)=([^&]*)/g;
+	let m: RegExpExecArray;
+	while ((m = regex.exec(fragmentString))) {
+		params[decodeURIComponent(m[1])] = decodeURIComponent(m[2]);
+	}
```
Comment (adirar111) [inline extension/ts/objects/auth.ts:95]: more weirdness taken from the docs

**Review by braxtonhall — APPROVED** [review]
approving my own changes :)

**Pull request merged**